boptimol/GP_MODEL.py

```diff
@@ -0,0 +1,239 @@
+    """
+    
+    mean_module = gpymeans.ConstantMean(batch_shape=torch.Size())
+    covar_module = gpykernels.ScaleKernel(gpykernels.RBFKernel())
+    
+    gp = SingleTaskGP(observed_X, observed_y, covar_module=covar_module, mean_module=mean_module)
+    mll = ExactMarginalLogLikelihood(gp.likelihood, gp).to(device=device)
+    fit_gpytorch_mll(mll)
+    
+    aq = LogProbabilityOfImprovement(gp, best_f=torch.max(observed_y))
+    candidate, acq_value = optimize_acqf(
+        aq, 
+        bounds=bounds,
+        q=1, 
+        num_restarts=64, 
+        raw_samples=64
+    )
+    return candidate.detach().numpy()[0, :]
+
+
+
+def select_next_point_dihedrals(bounds: torch.Tensor,
+        observed_X: List[List[float]], 
+        observed_y: List[float], 
+        device: str) -> np.ndarray:
+    
+    """Generate the next sample to evaluate
+
+    Uses BOTorch to pick the next sample using Expected Improvement
+
+    Args:
+        observed_X: Observed coordinates
+        observed_y: Observed energies
+    Returns:
+        Next coordinates to try
+    """
+    
+    mean_module = gpymeans.ConstantMean(batch_shape=torch.Size())
+    covar_module = gpykernels.ScaleKernel(gpykernels.ProductStructureKernel(
+                num_dims = observed_X.shape[1],
+                base_kernel=gpykernels.PeriodicKernel()))
+    
+    gp = SingleTaskGP(observed_X, observed_y, covar_module=covar_module, mean_module=mean_module)
+    mll = ExactMarginalLogLikelihood(gp.likelihood, gp).to(device=device)
+    fit_gpytorch_mll(mll)
+    
+    aq = LogExpectedImprovement(gp, best_f=torch.max(observed_y))
+    candidate, acq_value = optimize_acqf(
+        aq, 
+        bounds=bounds,
+        q=1, 
+        num_restarts=64, 
+        raw_samples=64
+    )
+    return candidate.detach().numpy()[0, :]
+    
+    
+
+def select_next_points_botorch( bounds: torch.Tensor,
+        observed_X: List[List[float]], 
+        observed_y: List[float],
+        mol: Molecule) -> np.ndarray:
+    
+    """Generate the next sample to evaluate
+
+    Uses BOTorch to pick the next sample using Expected Improvement
+
+    Args:
+        observed_X: Observed coordinates
+        observed_y: Observed energies
+    Returns:
+        Next coordinates to try
+    """
+
+    # Clipping the energies if needed
+    observed_y = np.clip(observed_y, -np.inf, 2 + np.log10(np.clip(observed_y, 1, np.inf)))
+
+    # Tracking the torch device
+    #  .. unfortuantely "MPS" for Apple Silicon doesn't support float64
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    # Convert inputs to torch arrays
+    train_X = torch.tensor(observed_X, dtype=torch.float64, device=device)
+    train_y = torch.tensor(observed_y, dtype=torch.float64, device=device)
+    
+    # Making this a maximization problem
+    train_Y = standardize(-1* train_y[:, None])
+
+    candidate_bond_lengths = select_next_point_bond_lengths(bounds[:, : mol.end_bonds],
```

Comment:
This definitely needs some comments - or at least assign the various train_X and bounds slices to named variables, e.g.
```
bond_bounds = bounds[:, : mol.end_bonds]
bond_X = train_X[:, : mol.end_bonds]
# …
angle_bounds = bounds[:, mol.end_bonds: mol.end_angles]
# …
```

It's important to think who will read the code .. which includes yourself in a few weeks.